File: _DEMO/agent-server/lib/agent.js
// Collections agent — tool definitions and agentic loop.

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { tools as TOOLS, callTool, getStatusMessage } from './mcp-client.js'; // [CHG-002]
import { callAiCore, streamResponse } from './ai-core.js';
import { log, err } from './logger.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
export const SYSTEM_PROMPT = fs.readFileSync(path.join(__dirname, 'system-prompt.txt'), 'utf8');

export function buildMessages(history, userMessage) {
  return [
    { role: 'system', content: SYSTEM_PROMPT },
    ...history.map(({ role, content }) => ({ role, content: String(content) })),
    { role: 'user', content: userMessage.trim() },
  ];
}

async function runToolCall(toolCall, send, context) {
  const { id, function: fn } = toolCall;
  const args = JSON.parse(fn.arguments || '{}');
  const statusMsg = getStatusMessage(fn.name);
  send.status(statusMsg);
  const start = Date.now();
  log('[mcp]', `→ ${fn.name}  "${statusMsg}"  args=${fn.arguments}`);
  let content;
  try {
    const result = await callTool(fn.name, args, context);
    log('[mcp]', `← ${fn.name}  ${Date.now() - start}ms  result=${result.slice(0, 120)}${result.length > 120 ? '…' : ''}`);
    content = JSON.stringify(result);
  } catch (e) {
    err('[mcp]', `← ${fn.name}  ${Date.now() - start}ms  ERROR: ${e.message}`);
    content = JSON.stringify({ error: `Tool ${fn.name} failed — please try again.` });
  }
  return { role: 'tool', tool_call_id: id, content };
}

export async function handleToolCalls(token, messages, choice, send, context = {}) {
  const toolCalls = choice.message.tool_calls;
  messages.push({ role: 'assistant', content: null, tool_calls: toolCalls });
  const toolResults = await Promise.all(toolCalls.map((tc) => runToolCall(tc, send, context)));
  for (const result of toolResults) messages.push(result);
  const t2start = Date.now();
  log('[ai-core]', `turn-2  streaming...`);
  const turn2Res = await callAiCore(token, { messages, stream: true });
  const chars = await streamResponse(turn2Res, send);
  log('[ai-core]', `turn-2  ${Date.now() - t2start}ms  chars=${chars}`);
}

File: _DEMO/agent-server-py/lib/agent.py
# [C03-F09] Agent logic — message building and agentic tool-call loop.

import asyncio
import json
import time
from pathlib import Path

from lib import mcp_client
from lib.ai_core import call_ai_core, stream_response
from lib.logger import log, err

_SYSTEM_PROMPT_PATH = Path(__file__).parent.parent.parent / "prompts" / "system-prompt.txt"
SYSTEM_PROMPT = _SYSTEM_PROMPT_PATH.read_text(encoding="utf-8")


def build_messages(history: list[dict], user_message: str) -> list[dict]:
    return [
        {"role": "system", "content": SYSTEM_PROMPT},
        *[{"role": t["role"], "content": str(t["content"])} for t in history],
        {"role": "user", "content": user_message.strip()},
    ]


async def _run_tool_call(tool_call: dict, send_status, context: dict) -> dict:
    tc_id = tool_call["id"]
    fn = tool_call["function"]
    args = json.loads(fn.get("arguments") or "{}")
    status_msg = mcp_client.get_status_message(fn["name"])
    await send_status(status_msg)
    start = time.monotonic()
    log("[mcp]", f"→ {fn['name']}  \"{status_msg}\"  args={fn.get('arguments')}")
    try:
        result = await mcp_client.call_tool(fn["name"], args, context)
        elapsed = int((time.monotonic() - start) * 1000)
        snippet = result[:120] + ("…" if len(result) > 120 else "")
        log("[mcp]", f"← {fn['name']}  {elapsed}ms  result={snippet}")
        content = json.dumps(result)
    except Exception as exc:
        elapsed = int((time.monotonic() - start) * 1000)
        err("[mcp]", f"← {fn['name']}  {elapsed}ms  ERROR: {exc}")
        content = json.dumps({"error": f"Tool {fn['name']} failed — please try again."})

    return {"role": "tool", "tool_call_id": tc_id, "content": content}


async def handle_tool_calls(token: str, messages: list[dict], choice: dict, send, context: dict = {}) -> None:
    tool_calls = choice["message"]["tool_calls"]
    messages.append({"role": "assistant", "content": None, "tool_calls": tool_calls})

    tool_results = await asyncio.gather(
        *[_run_tool_call(tc, send["status"], context) for tc in tool_calls]
    )
    for result in tool_results:
        messages.append(result)

    t2_start = time.monotonic()
    log("[ai-core]", "turn-2  streaming...")
    turn2_resp = await call_ai_core(token, {"messages": messages, "stream": True})

    async def _chunk(text: str) -> None:
        await send["chunk"](text)

    chars = await stream_response(turn2_resp, _chunk)
    elapsed = int((time.monotonic() - t2_start) * 1000)
    log("[ai-core]", f"turn-2  {elapsed}ms  chars={chars}")
